Add option `-L` to allow `docker cp` follow symbol link

WeiZhang555 opened the pull request:
fixes #16555

When user try to cp a symbol link to/from a container, what the user
really want is to copy the link target but not the symbol link, as the
symbol link is only valid and useful in host/container's scope.
All in all, who will need a symbol link copied to container which points 
to some file on host? That's meaningless.

And this is also the default behavior of a unix `cp` command.

Signed-off-by: Zhang Wei zhangwei555@huawei.com

WeiZhang555 commented:
`docker exec` test failed, I have no idea about this:
```
----------------------------------------------------------------------
FAIL: docker_cli_exec_test.go:615: DockerSuite.TestExecStartFails

docker_cli_exec_test.go:622:
    c.Fatal("exec error message not received. The daemon might had crashed")
... Error: exec error message not received. The daemon might had crashed


----------------------------------------------------------------------
```

calavera commented:
Removing a whole test without adding a new one makes me quite uncomfortable.

I'd like to know what @jlhawn thinks about this change before giving it my :+1:

WeiZhang555 commented:
@calavera Understand, that test case is wholy removed because it's totally inappropriate.
I'll add one more testcases when we make a deal about the `cp` behaviour. :)

WeiZhang555 commented:
@calavera 
Is there any way to run some specific integration test file instead of all? It's quite slow and inconvenient to run all the integration scripts. :)

tonistiigi commented:
Like pointed out in #16555 `docker cp` currently uses `cp -a` syntax. This isn't accidental, but documented, together with symlink following rules. We can consider supporting both, but probably not switching the behavior as we also have opposite issues from the time before refactoring took place(#13157).

More importantly, there is an API for pulling and pushing files to/from containers and there is a cli tool. IMHO if we allow following of the source we should do it in the client side and not change API. This is also how we do it currently when you copy local files into a symlink inside a container. We resolve the actual target directory in the client side and start upload to that.

WeiZhang555 commented:
@tonistiigi 
All right, how about that we reserve the default `docker cp` behaviour, but add more options such as `-L`(following symbol link) and `-r` (copy directory recursely), this will also need a little change to the API, e.g. `GET "/containers/{name:.*}/archive?recurse=true&linkFollow=false"`. 
Of course we can make default behaviour unchanged, but it's hard to change `cp` behaviour without modifying REST API. For example, the TARed file will contain file name, if we don't pass `linkFollow` param through to daemon, daemon can't choose a correct rebaseName.

What do you think of extra option of `-L` and `-r` ? Is that acceptable?

tonistiigi commented:
@WeiZhang555 I'm not sure if its better to add `-L` or make the current behavior `docker cp -a` but I'd really like if this would be client side change only. Take a look how we handle remote symlink targets. Also I think current rebasing all happens on client side anyway.

WeiZhang555 commented:
@tonistiigi 
We can make origal REST API `GET "/containers/{name:.*}/archive"` behave like `GET "/containers/{name:.*}/archive?recurse=true&linkFollow=false"`, so it won't impact exsiting tools. And IMO it's better. 

But it's also OK to change client only if you really feel uncomfortable with API change, I can't take a look at what I can do for that.

tonistiigi commented:
@WeiZhang555 I think the current API has the correct behavior. `/archive` does not assume any special handling for symlinks. We can reconsider if client-only change turns out to be impossible but I think its doable. Also make sure that copy-from-container and copy-to-container always work the same way.

WeiZhang555 commented:
@tonistiigi All right, I can try to find a way to make client-only change, maybe need a couple of days. :)

WeiZhang555 committed:
Add '-L' option for `cp`

Fixes #16555

Original docker `cp` always copy symbol link itself instead of target,
now we provide '-L' option to allow docker to follow symbol link to real
target.

Signed-off-by: Zhang Wei <zhangwei555@huawei.com>

WeiZhang555 commented:
Just refreshed my PR, add one '-L' option to allow following symbol links.
e.g.
`docker cp xxx:xxx xxx` means copy symbol link file itself;
`doker cp -L xxx:xxx xxx` means copy target file.

Let's see how it works.
I'm still working on test cases and docs thing.

WeiZhang555 commented:
@tonistiigi @calavera @jlhawn 
Just add some unit tests, I will keep working on integration test.
Please tell me what do you think of the `-L` option, is this OK for you?

One more thing, the reason why I use the `-L` instead of `-a` is that `-a`  is an option which is combined of `-d` and `-R`(reference unix `cp` command), but `-L` is kind of a standalone option, And also `-L` can keep original behavior of `docker cp` so it won't influence existing tools which makes use of `docker cp`.

Let me know whether this is acceptable for you in case  I  write too many codes that will never be merged. :)

tonistiigi commented:
@WeiZhang555 I didn't look at the code too carefully atm but the approach looks good to me. Tests look good as well but yes, inctegration-cli and docs are also needed. I wonder if `-R` would be a better flag because `-L` usually means resolve all symlinks but this is for resolving the source only.

tonistiigi commented on pkg/archive/copy_test.go:
```diff
@@ -310,11 +337,21 @@ func TestCopyCaseB(t *testing.T) {
-func TestCopyCaseC(t *testing.T) {
+// C.1 SRC specifies a file and DST exists as a file. This should overwrite
+//     the file at DST with the contents of the source file.
+func TestCopyCaseC1(t *testing.T) {
```
I'd use `TestCopyCaseCFollowSymlink` or `TestCopyCaseCFSym` instead the `1/2` suffix.

WeiZhang555 replied:
OK, I can fix this.

WeiZhang555 commented:
@tonistiigi We can find a nice option together,  but `-R` seems easy to be mixed with `-r`(recursively) and may not be a good choice. :)

I'll go on adding the integration testa and docs, also more tests to see if this change works correctly.  :)

WeiZhang555 commented:
@tonistiigi updated, please help review :)

vdemeester commented:
@WeiZhang555 sorry, needs a rebase :sweat:

WeiZhang555 commented:
@vdemeester 
Rebased :)

WeiZhang555 commented:
ping @tonistiigi ...
Any progress on this?

thaJeztah commented:
I think we're awaiting feedback from @jlhawn on this

ping @jlhawn could you have a look at this one pls?

WeiZhang555 commented:
> @tonistiigi
> We can make origal REST API GET "/containers/{name:.*}/archive" behave like GET "/containers/{name:.*}/archive?recurse=true&linkFollow=false", so it won't impact exsiting tools. And IMO it's better.

After some more considerations, I still think that we'd better to change the API. The current code only works for symlink to regular file but not directory, that's really not a perfect implementation. :disappointed: 

Without changing the `cp` API, it would be disaster to follow symbol link in a directory, because you have to download whole directory containing some symbol link files, then download all the linked target for all symlink iteratively (that may be hundreds or even thousands of requests), however this can be easily done on daemon side and still keep the backward compatibility if we do it correctly.

ping @all

tonistiigi commented:
@WeiZhang555 I think following subdirs/files is a different behavior than what we need atm. This PR is for adding the normal `cp` semantics compared to `cp -a` and is looking quite good for that. Only thing that is bit confusing is the flag but thats because we did `cp -a` without adding a flag. With some quick testing following symlink source directory seemed to work correctly(like `cp -r`), copying symlink for `symlink` and following it for `symlink/`.

WeiZhang555 commented:
It has been a while since last update, what do you think of this enhancement? :smile: 
ping @jlhawn

jlhawn commented:
Sorry for getting to this so late :-(

The behavior seems right. Thanks for the additional tests!

calavera commented:
LGTM

What do you think @tonistiigi? is it ready to move to docs review?

tonistiigi commented:
LGTM

LK4D4 commented:
ping @thaJeztah @moxiegirl

moxiegirl commented on docs/reference/commandline/cp.md:
```diff
@@ -16,7 +16,8 @@ weight=1
     Copy files/folders between a container and the local filesystem
 
-      --help=false        Print usage
+      -L, --follow-link=false    Always follow symbol link in source
```
Always follow symbolic links in source.

moxiegirl replied:
@WeiZhang555 not your fault but looking at the `cp` command it looks like the basic structure of the command description is flawed. If we restructure the command as a whole to use `CONTAINER:SRC/DEST_PATH` I think it is easier to follow the content.  Would you be willing to do this?

--> https://gist.github.com/moxiegirl/5b4cf0700075e343792a

# cp

    Usage: docker cp [OPTIONS] CONTAINER:SRC_PATH DEST_PATH | -
           docker cp [OPTIONS] SRC_PATH | - CONTAINER:DEST_PATH

    Copy files/folders between a container and the local filesystem

      -L, --follow-link=false    Always follow symbol link in SRC_PATH
      --help=false               Print usage

The `docker cp` utility copies the contents of `SRC_PATH` to the `DEST_PATH`.
You can copy from the container's file system to the local machine or the
reverse, from the local filesystem to the container. if If `-` is specified for
either the `SRC_PATH` or `DEST_PATH`, you can also stream a tar archive to
`STDOUT`. The `CONTAINER` can be a running or stopped container. The `SRC_PATH`
or  `DEST_PATH` be a file or directory.

The `docker cp` command assumes container paths are relative to the container's `/` (root) directory. This means supplying the initial forward slash is optional; The command sees `compassionate_darwin:/tmp/foo/myfile.txt` and
`compassionate_darwin:tmp/foo/myfile.txt` as identical. Local machine paths can
be an absolute or relative value. The command interprets a local machine's
relative paths as relative to the current working directory where `docker cp` is
run.

The `cp` command behaves like the Unix `cp -a` command in that directories are
copied recursively with permissions preserved if possible. Ownership is set to
the user and primary group at the destination. For example, files copied to a
container are created with `UID:GID` of the root user. Files copied to the local
machine are created with the `UID:GID` of the user which invoked the `docker cp`
command.  If you specify the `-L` option, `docker cp` follows any symbolic link
in the `SRC_PATH`.

Assuming a path separator of `/`, a first argument of `SRC_PATH` and second
argument of `DEST_PATH`, the behavior is as follows:

- `SRC_PATH` specifies a file
    - `DEST_PATH` does not exist
        - the file is saved to a file created at `DEST_PATH`
    - `DEST_PATH` does not exist and ends with `/`
        - Error condition: the destination directory must exist.
    - `DEST_PATH` exists and is a file
        - the destination is overwritten with the source file's contents
    - `DEST_PATH` exists and is a directory
        - the file is copied into this directory using the basename from
          `SRC_PATH`
- `SRC_PATH` specifies a directory
    - `DEST_PATH` does not exist
        - `DEST_PATH` is created as a directory and the *contents* of the source
           directory are copied into this directory
    - `DEST_PATH` exists and is a file
        - Error condition: cannot copy a directory to a file
    - `DEST_PATH` exists and is a directory
        - `SRC_PATH` does not end with `/.`
            - the source directory is copied into this directory
        - `SRC_PATH` does end with `/.`
            - the *content* of the source directory is copied into this
              directory

The command requires `SRC_PATH` and `DEST_PATH` to exist according to the above
rules. If `SRC_PATH` is local and is a symbolic link, the symbolic link, not
the target, is copied by default. To copy the link target and not the link, specify the `-L` option.

A colon (`:`) is used as a delimiter between `CONTAINER` and its path. You can
also use `:` when specifying paths to a `SRC_PATH` or `DEST_PATH` on a local
machine, for example  `file:name.txt`. If you use a `:` in a local machine path,
you must be explicit with a relative or absolute path, for example:

    `/path/to/file:name.txt` or `./file:name.txt`

It is not possible to copy certain system files such as resources under
`/proc`, `/sys`, `/dev`, and mounts created by the user in the container.

Using `-` as the `SRC_PATH` streams the contents of `STDIN` as a tar archive.
The command extracts the content of the tar to the `DEST_PATH` in container's
filesystem. In this case, `DEST_PATH` must specify a directory. Using `-` as
`DEST_PATH` streams the contents of the resource as a tar archive to `STDOUT`.

WeiZhang555 replied:
>@WeiZhang555 not your fault but looking at the cp command it looks like the basic structure >of the command description is flawed. If we restructure the command as a whole to use >CONTAINER:SRC/DEST_PATH I think it is easier to follow the content.
>
>Would you be willing to do this? If not, I can carry and ask you to do a technical review.

I'd like to help, always, so please let me do it if you have a good idea :smile: . 
But first of all, I couldn't quite get it. Do you mean we need a UX change? Can you explain a little bit more? thank you!

moxiegirl commented on docs/reference/commandline/cp.md:
```diff
@@ -34,12 +35,12 @@ supplying the initial forward slash is optional; The command sees
-files copied to a container will be created with `UID:GID` of the root user.
-Files copied to the local machine will be created with the `UID:GID` of the
-user which invoked the `docker cp` command.
+Behavior is similar to the common Unix utility `cp -a` (unless `-L` option is
```
The parenthetical throws doubt.  What happens if the command encounters a symbolic link. The comments in the code makes it sound like the command follows the link and copies it over with the rest of the contents from the source to the destination.  That's just a guess --> we need to describe it exactly.  If the link is to a file or to a directory does that make a diff?

```markdown
The `cp` command behaves like Unix utility `cp -a` in that directories are copied recursively with permissions preserved if possible. Ownership is set to the user and primary group at the destination. For example, files copied to a container are created with `UID:GID` of the root user. Files copied to the local machine are created with the `UID:GID` of the user which invoked the `docker cp` command.  If you specify the `-L` option, `docker cp` follows any symbolic link in the path...
```

---> preview

The `cp` command behaves like Unix utility `cp -a` in that directories are copied recursively with permissions preserved if possible. Ownership is set to the user and primary group at the destination. For example, files copied to a container are created with `UID:GID` of the root user. Files copied to the local machine are created with the `UID:GID` of the user which invoked the `docker cp` command.  If you specify the `-L` option, `docker cp` follows any symbolic link in the path...

moxiegirl commented on man/docker-cp.1.md:
```diff
@@ -28,12 +28,12 @@ supplying the initial forward slash is optional; The command sees
-files copied to a container will be created with `UID:GID` of the root user.
-Files copied to the local machine will be created with the `UID:GID` of the
-user which invoked the `docker cp` command.
+Behavior is similar to the common Unix utility `cp -a` (unless `-L` option is
```
@WeiZhang555 the man page should match whatever we decide to do with the `cp` reference

WeiZhang555 commented:
@moxiegirl Your suggestions of docs look good to me, I have included it in this commit already, please take a look to see if I miss something else or make something wrong.
Thank you  :smile:

moxiegirl commented on man/docker-cp.1.md:
```diff
@@ -147,6 +148,17 @@ to a directory `/etc/my-app.d` in a container:
 The above command will copy the contents of the local `/config` directory into
 the directory `/etc/my-app.d` in the container.
 
+Third, if you want to copy a symbol link into a container, what you want is to
```
@WeiZhang555 Wow! Thank you for doing that. Really cool.  

I missed this example in my review, it has some spelling and construction issues. Try this:

```markdown
Finally, if you want to copy a symbolic link into a container, you typically
want to  copy the linked target and not the link itself. To copy the target, use
the `-L` option, for example:

		$ ln -s /tmp/somefile /tmp/somefile.ln
		$ docker cp -L /tmp/somefile.ln myappcontainer:/tmp/

This command copies content of the local `/tmp/somefile` into the file
`/tmp/somefile.ln` in the container. Without `-L` option, the `/tmp/somefile.ln`
preserves its symbolic link but not its content.
```

WeiZhang555 replied:
@moxiegirl Thank you for help! :smile: 

Sorry that I just moved in another hotel with poor network connection, I'll update my PR once I get better internet connection later.

thaJeztah replied:
thanks, no worries @WeiZhang555

WeiZhang555 commented:
Updated.
Ping @moxiegirl @thaJeztah

thaJeztah commented on docs/reference/commandline/cp.md:
```diff
@@ -10,81 +10,76 @@ parent = "smn_cli"
-a tar archive to `STDOUT` if `-` is specified).
+The `docker cp` utility copies the contents of `SRC_PATH` to the `DEST_PATH`.
+You can copy from the container's file system to the local machine or the
+reverse, from the local filesystem to the container. if If `-` is specified for
```
double "if" here

WeiZhang555 replied:
Ah, you are right! :+1:

thaJeztah commented on docs/reference/commandline/cp.md:
```diff
@@ -10,81 +10,76 @@ parent = "smn_cli"
+reverse, from the local filesystem to the container. if If `-` is specified for
+either the `SRC_PATH` or `DEST_PATH`, you can also stream a tar archive to
+`STDOUT`. The `CONTAINER` can be a running or stopped container. The `SRC_PATH`
+or  `DEST_PATH` be a file or directory.
```
really minor nit; there's two spaces before `DEST_PATH` here :smile:

thaJeztah commented on docs/reference/commandline/cp.md:
```diff
@@ -10,81 +10,76 @@ parent = "smn_cli"
-In the second synopsis form, the contents of `LOCALPATH` (or a tar archive
-streamed from `STDIN` if `-` is specified) are copied from the local machine to
-`PATH` in the filesystem of `CONTAINER`.
+The `docker cp` command assumes container paths are relative to the container's `/` (root) directory. This means supplying the initial forward slash is optional; The command sees `compassionate_darwin:/tmp/foo/myfile.txt` and
```
can you wrap this line to 80 chars?

thaJeztah commented on man/docker-cp.1.md:
```diff
@@ -7,87 +7,88 @@ docker-cp - Copy files/folders between a container and the local filesystem.
+or  `DEST_PATH` be a file or directory.
 
-Behavior is similar to the common Unix utility `cp -a` in that directories are
+The `docker cp` command assumes container paths are relative to the container's `/` (root) directory. This means supplying the initial forward slash is optional; The command sees `compassionate_darwin:/tmp/foo/myfile.txt` and
```
can you wrap this line to 80 chars as well?

thaJeztah commented on man/docker-cp.1.md:
```diff
@@ -7,87 +7,88 @@ docker-cp - Copy files/folders between a container and the local filesystem.
 
 In the first synopsis form, the `docker cp` utility copies the contents of
-`PATH` from the filesystem of `CONTAINER` to the `LOCALPATH` (or stream as
+`SRC_PATH` from the filesystem of `CONTAINER` to the `DEST_PATH` (or stream as
```
The general description in the man page is different from the `docs/...` page, I think @moxiegirl meant to make these the same

@moxiegirl am I correct that you wanted that?

thaJeztah commented on man/docker-cp.1.md:
```diff
@@ -147,6 +148,17 @@ to a directory `/etc/my-app.d` in a container:
+the `-L` option, for example:
+
+		$ ln -s /tmp/somefile /tmp/somefile.ln
+		$ docker cp -L /tmp/somefile.ln myappcontainer:/tmp/
```
Looks like these lines are indented one tab too much. Also, could you change to 4 spaces in stead of tabs?

thaJeztah commented:
Thanks @WeiZhang555, I added my comments inline

WeiZhang555 commented:
@thaJeztah Very detailed suggestion, I'll do it now :+1:

WeiZhang555 commented:
@thaJeztah @moxiegirl updated.

>If `-` is specified for
>either the `SRC_PATH` or `DEST_PATH`, you can also stream a tar archive to
>`STDOUT`.

I think it's not very accurate. How about
```
If `-` is specified for
either the `SRC_PATH` or `DEST_PATH`, you can also stream a tar archive from `STDIN` or to
`STDOUT`.
```

Maybe this is better? WDYT?

thaJeztah commented:
Hm, yes, I think that makes sense

WeiZhang555 commented:
@thaJeztah updated. :smile:

moxiegirl commented:
LGTM! Thank you @WeiZhang555 :-D

thaJeztah commented:
Thanks @WeiZhang555 for your contribution!

LGTM

thaJeztah commented:
restarting Windows build

vdemeester commented:
Windows builds are definitely in holidays.. merging :sweat_smile:

WeiZhang555 commented:
Yeah, Windows build is a naughty boy :smile: 

Thank you all for helping!  :laughing:
